src/lib.rs

```diff
@@ -424,37 +456,58 @@ async fn start(
                 config.blocks_preload_pool_size,
                 streamer_messages_futures.len(),
             );
-
-            let blocks_preload_pool_current_len = streamer_messages_futures.len();
-            streamer_messages_futures.extend(
-                fast_fetch_block_heights(
-                    &mut pending_block_heights,
-                    config
-                        .blocks_preload_pool_size
-                        .saturating_sub(blocks_preload_pool_current_len),
-                    blocks_preload_pool_current_len == 0,
-                )
-                .await?
-                .into_iter()
-                .map(|block_height| {
-                    s3_fetchers::fetch_streamer_message(
-                        &s3_client,
-                        &config.s3_bucket_name,
-                        block_height,
-                    )
-                }),
-            );
-
             tracing::debug!(
                 target: LAKE_FRAMEWORK,
                 "Streaming block #{} ({})",
                 streamer_message.block.header.height,
                 streamer_message.block.header.hash
             );
-            if let Err(SendError(_)) = streamer_message_sink.send(streamer_message).await {
-                tracing::debug!(target: LAKE_FRAMEWORK, "Channel closed, exiting");
+            let blocks_preload_pool_current_len = streamer_messages_futures.len();
+
+            let prefetched_block_heights_future = prefetch_block_heights_into_pool(
+                &mut pending_block_heights,
+                config
+                    .blocks_preload_pool_size
+                    .saturating_sub(blocks_preload_pool_current_len),
+                blocks_preload_pool_current_len == 0,
+            );
+
+            let streamer_message_sink_send_future = streamer_message_sink.send(streamer_message);
+
+            let (prefetch_res, send_res): (
+                Result<Vec<types::BlockHeight>, anyhow::Error>,
+                Result<_, SendError<near_indexer_primitives::StreamerMessage>>,
+            ) = futures::join!(
+                prefetched_block_heights_future,
+                streamer_message_sink_send_future,
+            );
+
+            if let Err(SendError(err)) = send_res {
+                tracing::debug!(
+                    target: LAKE_FRAMEWORK,
+                    "Failed to send StreamerMessage (#{:0>12}) to the channel. Channel is closed, exiting \n{:?}",
+                    start_from_block_height - 1,
+                    err,
+                );
                 return Ok(());
             }
+
+            streamer_messages_futures.extend(prefetch_res?.into_iter().map(|block_height| {
```

Comment:
By the way, `?` in this line can lead us to a similar no-log silent error case. Let's add `map_err()` which will log and return an error, and only then use `?`. Note, apply it to the other places with the question mark operator in this function.